core/sched/rt-rm.c

```diff
@@ -0,0 +1,233 @@
+#include <core/sched/scheduler_skeleton.h>
+#include <core/sched/sched-config.h>
+#include <config.h>
+#include <arch/armv7.h>
+#include <stdlib.h>
+#include <debug.h>
+#include <stdbool.h>
+
+#include <lib/list.h>
+#include <core/timer.h>
+
+// #define SCHED_RM_REPORT
+
+struct entry_data_rm {
+    struct sched_entry *e;
+
+    uint32_t period;
+    uint32_t budget;
+
+    uint32_t period_cntdn;
+    uint32_t budget_cntdn;
+
+#ifdef SCHED_RM_REPORT
+    /* for analysis */
+    uint32_t foreground_ticks;
+    uint32_t preempted;
+#endif
+
+    struct list_head head;
+};
+
+struct sched_data_rm {
+    struct scheduler *s;
+
+    uint64_t tick_interval_us;
+    uint64_t tick_count;
+#ifdef SCHED_RM_REPORT
+    uint64_t tick_1kcount;
+#endif
+
+    struct entry_data_rm *current;
+    struct list_head runqueue;
+};
+
+/* Scheduler related data initialization */
+void sched_rm_init(struct scheduler *s)
+{
+    struct sched_data_rm *sd = (struct sched_data_rm *) malloc(sizeof(struct sched_data_rm));
+
+    /* Check scheduler config */
+    /* Allocate memory for system-wide data */
+    /* Initialize data */
+    sd->current = NULL;
+    sd->tick_count = 0;
+#ifdef SCHED_RM_REPORT
+    sd->tick_1kcount = 1000;
+#endif
+    sd->tick_interval_us = schedconf_rm_tick_interval_ms[s->pcpuid];
+    sd->s = s;
+
+    LIST_INITHEAD(&sd->runqueue);
+
+    s->sd = sd;
+}
+
+int sched_rm_vcpu_register(struct scheduler *s, struct sched_entry *e)
+{
+    // struct sched_data_rm *sd = (struct sched_data_rm *) (s->sd);
+    struct entry_data_rm *ed = (struct entry_data_rm *) malloc(sizeof(struct entry_data_rm));
+
+    ed->period = schedconf_rm_period_budget[e->vcpuid][0];
+    ed->budget = schedconf_rm_period_budget[e->vcpuid][1];
+    ed->e = e;
+
+#ifdef SCHED_RM_REPORT
+    ed->foreground_ticks = 0;
+    ed->preempted = 0;
+#endif
+
+    LIST_INITHEAD(&ed->head);
+
+    e->ed = ed;
+
+    return 0;
+}
+
+int sched_rm_vcpu_unregister(struct scheduler *s, struct sched_entry *e)
+{
+    /* Check if vcpu is registered */
+    /* Check if vcpu is detached. If not, request detachment.*/
+    /* If we have requested detachment of vcpu,
+     *   let's wait until it is detached by main scheduling routine */
+
+    return 0;
+}
+
+int sched_rm_vcpu_attach(struct scheduler *s, struct sched_entry *e)
+{
+    struct sched_data_rm *sd = (struct sched_data_rm *) (s->sd);
+    struct entry_data_rm *ed = (struct entry_data_rm *) (e->ed);
+
+    ed->period_cntdn = ed->period;
```

Comment:
When we use variable for countdown, we may expect that the variable is subtracted.
But L102 and L103 do not work like expectation, let me know why and don't we have other approach?